File: soln-grid-problems/DP-maximum-size-sub-matrix-with-all-1s-in-a-binary-matrix.java
/*
https://www.geeksforgeeks.org/maximum-size-sub-matrix-with-all-1s-in-a-binary-matrix/
https://leetcode.com/problems/maximal-square/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public int maximalSquare(char[][] matrix) {
        int dp[][]=new int[matrix.length+1][matrix[0].length+1];
        int max=0;
        for(int i=1;i<dp.length;i++){
            for(int j=1;j<dp[0].length;j++){
                if(matrix[i-1][j-1]=='0'){
                    dp[i][j]=0;
                }
                else{
                    dp[i][j]=Math.min(Math.min(dp[i][j-1],dp[i-1][j]),dp[i-1][j-1])+1;
                    max=Math.max(max,dp[i][j]);
                }
            }
        }
        return max*max;
    }
}

File: C++/soln-grid-problems/DP-maximum-size-sub-matrix-with-all-1s-in-a-binary-matrix.cpp
  
/**
 @author Farheen Bano

 Date 24-07-2021   

 References-
 https://www.geeksforgeeks.org/maximum-size-sub-matrix-with-all-1s-in-a-binary-matrix/
 https://leetcode.com/problems/maximal-square/
*/

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int n=matrix.size(),m=matrix[0].size();
        vector<vector<int>> newMatrix(n+1, vector<int>(m+1,0));
        int maxSq=0;
        for(int i=1;i<n+1;i++){
            for(int j=1;j<m+1;j++){
                if(matrix[i-1][j-1]=='0')
                    newMatrix[i][j]=0;
                else{
                    newMatrix[i][j]=1+min(
                        min(
                            newMatrix[i][j-1],
                            newMatrix[i-1][j]                            
                        ),
                        newMatrix[i-1][j-1]
                    );
                    maxSq=max(maxSq,newMatrix[i][j]);
                        
                }
            }
        }
        return maxSq*maxSq; 
    }
};